core/trino-main/src/main/java/io/trino/operator/GroupByIdBlock.java

```diff
@@ -13,31 +13,76 @@
  */
 package io.trino.operator;
 
+import com.google.common.annotations.VisibleForTesting;
 import io.trino.spi.block.Block;
 import org.openjdk.jol.info.ClassLayout;
 
+import java.util.Arrays;
 import java.util.List;
 import java.util.OptionalInt;
 import java.util.function.ObjLongConsumer;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
-import static io.trino.spi.type.BigintType.BIGINT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
 import static java.util.Collections.singletonList;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * A block holding 'long' (8-byte) values and an additional group count value.
+ * The implementation covers both standard and rle encoding through a simple
+ * trick of a logical conjunction (&) of the position with a mask that is
+ * full (111...111) for a standard encoding and empty (000...000) for a rle.
+ * This way the rle block will always return the first position from the underlying array.
+ * <p>
+ * The implementation guarantees that every fetch of a single value will not
+ * result in a virtual call and, as a result, may be safely inlined by the compiler.
+ * <p>
+ * The only limitation is that the block cannot be dictionary encoded, but this encoding
+ * is not used within group by logic.
+ */
 public class GroupByIdBlock
         implements Block
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupByIdBlock.class).instanceSize();
+    private static final int SIZE_OF_ELEMENT = Long.BYTES;
+
+    private static final int MASK_STANDARD = -1;
+    private static final int MASK_RLE = 0;
 
     private final long groupCount;
-    private final Block block;
+    private final int positionCount;
+    private final long[] groups;
+    private final int mask;
+
+    public static GroupByIdBlock standard(long groupCount, long[] groupIds)
+    {
+        return new GroupByIdBlock(groupCount, groupIds.length, groupIds, MASK_STANDARD);
+    }
+
+    public static GroupByIdBlock rle(long groupCount, long value, int length)
+    {
+        return new GroupByIdBlock(groupCount, length, new long[] {value}, MASK_RLE);
+    }
+
+    public static GroupByIdBlock ofBlock(long groupCount, Block block)
+    {
+        if (block instanceof GroupByIdBlock && ((GroupByIdBlock) block).getGroupCount() == groupCount) {
```

Comment:
Can it really have a different group count if `block` is `GroupByIdBlock`? Could it have more groups?

It kind of seems this is some very specific aggregation logic (distinct, ordered) which maybe needs to be implemented there

Reply:
ping

Reply:
Why not? Why do we have to consider this? The method is prepared for all possible input arguments

Reply:
So this condition `if (block instanceof GroupByIdBlock && ((GroupByIdBlock) block).getGroupCount() == groupCount)` is not dead code?

Reply:
Actually this code branch is executed only for blocks that are not `GroupByIdBlock` so we can probably just get rid of the whole branch altogether.